File: go/bob/bob.go
package bob // package name must match the package name in bob_test.go
import "strings"

const testVersion = 2 // same as targetTestVersion
func Hey(s string) string {
	if AllUpper(s) {
		return "Whoa, chill out!"
	} 
        if EndWithQuestion(strings.TrimSpace(s)) {
		return "Sure."
	} 
        if AllNonCharacter(strings.TrimSpace(s)) {
		return "Fine. Be that way!"
	} 
        return "Whatever."
}

func EndWithQuestion(s string) bool {
	return len(s) > 0 && s[len(s)-1] == '?'
}

func AllUpper(s string) bool {
	containNonUpper := false
	containUpper := false
	for _, c := range s {
		if (c >= 'A' && c <= 'Z') || (c == '!') {
			containUpper = true
		} else if c >= 'a' && c <= 'z' {
			containNonUpper = true
		}
	}
	return containUpper && !containNonUpper && (strings.ToUpper(s) == s)
}

func AllNonCharacter(s string) bool {
	if len(s) == 0 {
		return true
	}
	for _, c := range s {
		if (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') {
			return false
		}
	}
	return false
}

File: haskell/bob/src/Bob.hs
module Bob (responseFor) where

responseFor :: String -> String
responseFor xs
  | length(trim xs) == 0 = "Fine. Be that way!"
  | onlyNumber xs = "Whatever."
  | allUpper xs = "Whoa, chill out!"
  | endWithQuestion(trim xs) = "Sure."
  | allNonCharacter xs = "Fine. Be that way!"
  | otherwise = "Whatever."
  where trim = unwords . words

onlyNumber :: String -> Bool
onlyNumber [] = True
onlyNumber (x:xs)
  | x == '!' || x == ' ' || x == ',' || x >= '0' && x <= '9' = onlyNumber xs
  | otherwise = False

containUpper :: String -> Bool
containUpper [] = False
containUpper (x:xs)
  | x >= 'A' && x <= 'Z' || x == '!' = True
  | otherwise = containUpper xs

containNonUpper :: String -> Bool
containNonUpper [] = False
containNonUpper (x:xs)
  | x >= 'a' && x <= 'z' = True
  | otherwise = containNonUpper xs

allUpper :: String -> Bool
allUpper xs = (containUpper xs) && not (containNonUpper xs)

endWithQuestion :: String -> Bool
endWithQuestion [] = False
endWithQuestion "?" = True
endWithQuestion (x:xs) = endWithQuestion xs

endWithExclamation :: String -> Bool
endWithExclamation [] = False
endWithExclamation "!" = True
endWithExclamation (x:xs) = endWithExclamation xs

allNonCharacter :: String -> Bool
allNonCharacter xs
  | length xs == 0 = True
allNonCharacter [] = True
allNonCharacter (x:xs)
  | x >= 'a' && x <= 'z' || x >= 'A' && x <= 'B' = False
  | otherwise = allNonCharacter xs
